bundles/auth/org.eclipse.smarthome.auth.oauth2client/src/main/java/org/eclipse/smarthome/auth/oauth2client/internal/cipher/SymmetricKeyCipher.java

```diff
@@ -0,0 +1,176 @@
+        if (plainText == null) {
+            return null;
+        }
+
+        // Generate IV
+        byte iv[] = new byte[IV_BYTE_SIZE];
+        random.nextBytes(iv);
+        cipherEnc = Cipher.getInstance(ENCRYPTION_ALGO);
+        cipherEnc.init(Cipher.ENCRYPT_MODE, encryptionKey, new IvParameterSpec(iv));
+        byte[] encryptedBytes = cipherEnc.doFinal(plainText.getBytes(StandardCharsets.UTF_8));
+        byte[] encryptedBytesWithIV = new byte[encryptedBytes.length + IV_BYTE_SIZE];
+
+        // copy iv to the start of array
+        System.arraycopy(iv, 0, encryptedBytesWithIV, 0, IV_BYTE_SIZE);
+        // append encrypted text to tail
+        System.arraycopy(encryptedBytes, 0, encryptedBytesWithIV, IV_BYTE_SIZE, encryptedBytes.length);
+        String encryptedBase64String = Base64.getEncoder().encodeToString(encryptedBytesWithIV);
+
+        return encryptedBase64String;
+    }
+
+    @Override
+    public String decrypt(@Nullable String base64CipherText) throws GeneralSecurityException {
+        if (base64CipherText == null) {
+            return null;
+        }
+        // base64 decode the base64CipherText
+        byte[] decodedCipherTextWithIV = Base64.getDecoder().decode(base64CipherText);
+        // Read IV
+        byte[] iv = new byte[IV_BYTE_SIZE];
+        System.arraycopy(decodedCipherTextWithIV, 0, iv, 0, IV_BYTE_SIZE);
+
+        byte[] cipherTextBytes = new byte[decodedCipherTextWithIV.length - IV_BYTE_SIZE];
+        System.arraycopy(decodedCipherTextWithIV, IV_BYTE_SIZE, cipherTextBytes, 0, cipherTextBytes.length);
+
+        Cipher cipherDec = Cipher.getInstance(ENCRYPTION_ALGO);
+        cipherDec.init(Cipher.DECRYPT_MODE, encryptionKey, new IvParameterSpec(iv));
+        byte[] decryptedBytes = cipherDec.doFinal(cipherTextBytes);
+        return new String(decryptedBytes, StandardCharsets.UTF_8);
+    }
+
+    private static SecretKey generateEncryptionKey() throws NoSuchAlgorithmException {
+        KeyGenerator keygen = KeyGenerator.getInstance("AES");
+        keygen.init(128); // do not use very high grade encryption due to export limit
+        SecretKey secretKey = keygen.generateKey();
+        return secretKey;
+    }
+
+    private SecretKey getOrGenerateEncryptionKey(Configuration configuration) {
+        String encryptionKeyInBase64 = null;
+        Dictionary<String, Object> properties = configuration.getProperties();
+        if (properties == null) {
+            properties = new Hashtable<String, Object>();
+        }
+
+        if (properties.get(PROPERTY_KEY_ENCRYPTION_KEY_BASE64) == null) {
+            try {
+                encryptionKey = generateEncryptionKey();
+                encryptionKeyInBase64 = new String(Base64.getEncoder().encode(encryptionKey.getEncoded()));
+
+                // Put encryption key back into config
+                properties.put(PROPERTY_KEY_ENCRYPTION_KEY_BASE64, encryptionKeyInBase64);
+                properties.put("service.pid", PID);
```

Comment:
You don't need to add service pid property. You just need to change the property that is expected to be updated.